app/com/microsoft/azure/iotsolutions/devicetelemetry/services/models/EmailServiceModel.java

```
@@ -0,0 +1,88 @@
+package com.microsoft.azure.iotsolutions.devicetelemetry.services.models;
+
+import com.fasterxml.jackson.databind.PropertyNamingStrategy;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonNaming;
+import com.microsoft.azure.iotsolutions.devicetelemetry.services.exceptions.InvalidInputException;
+
+import javax.mail.internet.AddressException;
+import javax.mail.internet.InternetAddress;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+@JsonDeserialize(as = EmailServiceModel.class)
+@JsonNaming(PropertyNamingStrategy.UpperCamelCaseStrategy.class)
+public final class EmailServiceModel implements IActionServiceModel {
+
+    private Type Type = IActionServiceModel.Type.Email;
+    public String Subject = "";
+    public String Body = "";
+    public List<String> Email = new ArrayList<>();
+
+    public EmailServiceModel() {
+        //empty constructor
+    }
+
+    public Type getType() {
+        return Type;
+    }
+
+    public void setType(Type Type) {
+        this.Type = Type;
+    }
+
+    public EmailServiceModel(IActionServiceModel.Type type, Map<String, Object> parameters) throws InvalidInputException {
+        Type = type;
+        if (parameters.containsKey("Subject")) {
+            Subject = (String) parameters.get("Subject");
+        }
+        if (parameters.containsKey("Template")) {
+            Body = (String) parameters.get("Template");
+        }
+
+        try {
+            if (parameters.containsKey("Email")) {
+                try {
+                    Email = (ArrayList<String>) parameters.get("Email");
+                } catch (ClassCastException e) {
+                    throw new InvalidInputException("Email field is a list of string");
+                }
+            }
+        } catch (Exception e) {
+            throw new InvalidInputException("Invalid input");
+        }
+
+        if (!isValid()) {
```

I would move the exception throwing here to the isValid method since it is already throwing exceptions most of the time.

---

Fixed

---

I would also throw an invalid input exception from isValid instead of returning false